### ISSUE TITLE
Aggregated Indexes for all!

### ISSUE THREAD

**Issue by goldmerc** [laravel/scout#599]
Hi, I'm a big fan of algolia scout extended's aggregator feature and it seems to me that the concept of an aggregated index that resolves to eloquent models of various different classes could be engine agnostic and a really useful feature for all the different scout engines. 

I would love to see aggregate indexes become part of the main scout package. Sadly, coding is not my day job and this would be above my pay grade but i'd be happy to write documentation or help out any way I can.

Anyway, just a thought! Thanks for all the amazing work you guys all do!

**Comment by driesvints** [comment]
I do agree that some sort of aggregated index would be useful for Scout. However, making it work on all different engines will be quite the challenge I believe. I'll add this internal backlog to maybe tackle some day but for now we don't have time to put into it ourselves. We'd maybe be open to a PR to kickstart it though. Thanks

**Comment by goldmerc** [comment]
Thanks @driesvints. I don’t think my coding is up to the level of submitting a PR for this but imagining for a second that it was, I’m wondering about how it could work.

I played a little the last few days and it's easy to make an aggregate index using the same return value for searchableAs() on multiple models and then having getScoutKey() return a string with the object class and ID. Then you just need to change how the model objects are resolved from the returned objectIDs and you’re in business. In my case, I had two models sharing a parent, so I could just override getScoutModelsByIds() in the parent…

```
public function getScoutModelsByIds($builder, $ids)
{
    return collect($ids)->mapToGroups(function ($item) {
        list($class, $id) = explode('::', $item);
        return [$class => $id];
    })->map(function ($ids, $class) {
        return $class::find($ids);
    })->collapse();
}
```

A quick hack but sufficient for what I was doing and seems to work.

The problem with an approach like this is that you then don’t have the ability to have individual scout indexes for the models you are aggregating. The way algolia/scout-extended works is that each aggregate gets it’s own searchable class. Anyway, the point of all this is that I got to wondering why Laravel/Scout doesn’t create an independent searchable class to handle the scout indexes for each model? Is there not any use case when you might want to have multiple indexes related to a single model? For performance or security reasons perhaps? If Scout did use separate searchable classes, I think it would be easier and cleaner to implement aggregate indexes in future. 

Anyhoo, as I said, I’m probably not going to be able to take this any further and I’m just chatting really, so feel free to ignore me as I’m sure you’re busy doing real work! ;-)
